src/main/java/org/truffleruby/core/support/RubyRandomizer.java

```diff
@@ -18,10 +18,12 @@
 public final class RubyRandomizer extends RubyDynamicObject {
 
     public Randomizer randomizer;
+    public final boolean threadSafe; // Used to configure new Randomizer instances, e.g. when setting seed manually
```

Comment:
I think it would be more elegant to make `Randomizer.seed` non-`final`, there seems to be very little advantages for it to be `final`, and that would avoid keeping `threadSafe` in two places.
To do that, we'd add Randomizer#setSeed() and that would also reset the `state` like constructors.

Could you try that?

Reply:
I tried this path but wasn't entirely happy with the result, since it made `Randomizer` and the code calling it in `RandomizerNodes` more complex. (`Randomizer` spent a lot more code in initialization, and `RandomizerNodes` needed a a few more changes because `randomFromBignum` is used both to initialize a new `RubyRandomizer` as well as updating the seed).

Instead, I've gone with simplifying the `Randomizer` and moving the synchronization entirely to `RubyRandomizer`. Let me know what you think about this variant! :)

Reply:
That looks fine.
It does synchronize the `Tempering` logic though, which could increase contention, but anyway if people want to avoid contention they shouldn't use a Random instance from multiple threads.
Only concern is Randomizer#genrandInt32 shouldn't be called by anything except RubyRandomizer.
Maybe by renaming it to unsynchronizedGenRandInt32()? Could you do that?

I think eventually there is no good reason to have Randomizer and RubyRandomizer as separate classes, so probably a good option is to merge Randomizer in RubyRandomizer. That seems out of scope of this PR though.

Reply:
> It does synchronize the Tempering logic though, which could increase contention [...]

Indeed, but since it's all pure integer math on a single register and using a few constants it should be by only a few extra cpu cycles (once optimized), so I expect the cost of the memory barriers and whatnot from the lock will completely dwarf it.

> Maybe by renaming it to unsynchronizedGenRandInt32()?

:+1: Done